source/common/statistic_impl.h

```diff
@@ -150,4 +151,124 @@ class HdrStatistic : public StatisticImpl {
+        worker_id_(worker_id) {}
+
+  ~SinkableStatistic() override {
+    // We must explicitly free the StatName here in order to supply the
+    // SymbolTable reference.
+    MetricImpl::clear(scope_.symbolTable());
+  }
+
+  // Currently Envoy Histogram Unit supports {Unspecified, Bytes, Microseconds, Milliseconds}. By
+  // default, Nighthawk::Statistic uses nanosecond as the unit of latency histograms, so Unspecified
+  // is returned here to isolate Nighthawk Statistic from Envoy Histogram Unit.
+  Envoy::Stats::Histogram::Unit unit() const override {
+    return Envoy::Stats::Histogram::Unit::Unspecified;
+  };
+  Envoy::Stats::SymbolTable& symbolTable() override { return scope_.symbolTable(); }
+
+  const absl::optional<int> worker_id() { return worker_id_; }
+
+protected:
+  // This is used for delivering the histogram data to sinks.
+  Envoy::Stats::Scope& scope_;
+
+private:
+  // worker_id can be used in downstream stats Sinks as the stats tag.
+  absl::optional<int> worker_id_;
+};
+
+// Implementation of sinkable Nighthawk Statistic with HdrHistogram.
+class SinkableHdrStatistic : public SinkableStatistic, public HdrStatistic {
+public:
+  SinkableHdrStatistic(Envoy::Stats::Scope& scope,
+                       const absl::optional<int> worker_id = absl::nullopt)
+      : SinkableStatistic(scope, worker_id) {}
+
+  // Envoy::Stats::Histogram
+  void recordValue(uint64_t value) override {
+    addValue(value);
+    // Currently in Envoy Scope implementation, deliverHistogramToSinks() will flush the histogram
+    // value directly to stats Sinks.
+    scope_.deliverHistogramToSinks(*this, value);
+  }
+  bool used() const override { return count() > 0; }
+  // Overriding name() to return Nighthawk::Statistic::id().
+  std::string name() const override { return id(); }
+  std::string tagExtractedName() const override {
+    RELEASE_ASSERT(false, "tagExtractedName() should not be called in Nighthawk Statistic");
+    return id();
+  }
+};
+
+// Implementation of sinkable Nighthawk Statistic with Circllhist Histogram.
+class SinkableCircllhistStatistic : public SinkableStatistic, public CircllhistStatistic {
+public:
+  SinkableCircllhistStatistic(Envoy::Stats::Scope& scope,
+                              const absl::optional<int> worker_id = absl::nullopt)
+      : SinkableStatistic(scope, worker_id) {}
+
+  // Envoy::Stats::Histogram
+  void recordValue(uint64_t value) override {
+    addValue(value);
+    // Currently in Envoy Scope implementation, deliverHistogramToSinks() will flush the histogram
+    // value directly to stats Sinks.
+    scope_.deliverHistogramToSinks(*this, value);
+  }
+  bool used() const override { return count() > 0; }
+  // Overriding name() to return Nighthawk::Statistic::id().
+  std::string name() const override { return id(); }
+  std::string tagExtractedName() const override {
+    RELEASE_ASSERT(false, "tagExtractedName() should not be called in Nighthawk Statistic");
```

Comment:
Sorry -- maybe we can replace this and the return with `NOT_IMPLEMENTED_GCOVR_EXCL_LINE`, e.g.

```c++
std::string tagExtractedName() const override { NOT_IMPLEMENTED_GCOVR_EXCL_LINE; }
```

This expresses that we don't expect this to be called, and should assert when that does happen.
It may be wishful thinking, but hopefully someday our coverage will actually also exclude these lines from its report.

Reply:
Thanks Otto. Changed to NOT_IMPLEMENTED_GCOVR_EXCL_LINE.

Also lower the coverage threshold from 98.6 to 98.4 since there is no meaningful way to test the lines that are missing coverage.

Reply:
makes sense, thanks